DllImportGenerator/DllImportGenerator/Marshalling/StringMarshaller.Utf8.cs

```diff
@@ -0,0 +1,280 @@
+                    yield return LocalDeclarationStatement(
+                        VariableDeclaration(
+                            AsNativeType(info),
+                            SingletonSeparatedList(VariableDeclarator(nativeIdentifier))));
+                    break;
+                case StubCodeContext.Stage.Marshal:
+                    if (info.RefKind != RefKind.Out)
+                    {
+                        // <native> = (byte*)Marshal.StringToCoTaskMemUTF8(<managed>);
+                        var coTaskMemAlloc = ExpressionStatement(
+                            AssignmentExpression(
+                                SyntaxKind.SimpleAssignmentExpression,
+                                IdentifierName(nativeIdentifier),
+                                CastExpression(
+                                    NativeType,
+                                    InvocationExpression(
+                                        MemberAccessExpression(
+                                            SyntaxKind.SimpleMemberAccessExpression,
+                                            InteropServicesMarshalType,
+                                            IdentifierName("StringToCoTaskMemUTF8")),
+                                        ArgumentList(
+                                            SingletonSeparatedList<ArgumentSyntax>(
+                                                Argument(IdentifierName(managedIdentifier))))))));
+                        if (info.IsByRef && info.RefKind != RefKind.In)
+                        {
+                            yield return coTaskMemAlloc;
+                        }
+                        else
+                        {
+                            // <usedCoTaskMemIdentifier> = false;
+                            yield return LocalDeclarationStatement(
+                                VariableDeclaration(
+                                    PredefinedType(Token(SyntaxKind.BoolKeyword)),
+                                    SingletonSeparatedList(
+                                        VariableDeclarator(usedCoTaskMemIdentifier)
+                                            .WithInitializer(EqualsValueClause(LiteralExpression(SyntaxKind.FalseLiteralExpression))))));
+
+                            string stackAllocPtrIdentifier = $"{managedIdentifier}__stackalloc";
+                            string fixedCharIdentifier = $"{managedIdentifier}__fixedChar";
+                            string byteLenIdentifier = $"{managedIdentifier}__byteLen";
+
+                            // Code block for stackalloc if string is below threshold size
+                            var marshalOnStack = Block(
+                                // byte* <stackAllocPtr> = stackalloc byte[<byteLen>];
+                                LocalDeclarationStatement(
+                                    VariableDeclaration(
+                                        PointerType(PredefinedType(Token(SyntaxKind.ByteKeyword))),
```

Comment:
I think if we use the `ReadOnlySpan<char>, Span<byte>` override for `GetBytes`, the code will be easier to read and won't require us to pin the string. We'll have to do some conversion for the argument from a byref to a pointer, but that's pretty simple to do at that point.

(You can assign a stackalloc to a `Span<T>` without a cast. Roslyn generates the constructor call for you with the correct length).

Reply:
Good call on no string pinning.

I can keep assigning the stackalloc to a pointer and do:
```
byte* nativePtr;
...
byte* stackAllocPtr = stackalloc byte[byteLen];
Encoding.UTF8.GetBytes(str.AsSpan(), new Span<byte>(stackAllocPtr, byteLen));
nativePtr = stackAllocPtr;
```
Which I think would be equivalent to what you are saying.

Maybe I'm being dense / don't know how to use spans, but if I assign the stackalloc to a `Span`, how do get back at the pointer reasonaably? I know there's `fixed ( ... = &MemoryMarshal.GetReference(...))`, but that seems like extra complication compared to just keeping it as a pointer and constructing the span explicitly?

Reply:
Your way with saving the pointer and then constructing the Span would work. I was thinking of using `Unsafe.AsPointer(span[0])` to get the pointer (which since it’s a stackalloc doesn’t need to be pinned). I’m fine if we don’t want to do that though.